symengine/refine.cpp

```diff
@@ -70,6 +70,78 @@ void RefineVisitor::bvisit(const Conjugate &x)
     }
 }
 
+void RefineVisitor::bvisit(const Max &x)
+{
+    // positive > nonpositive
+    // nonnegative and positive > negative
+    vec_basic nonpositive;
+    vec_basic negative;
+    vec_basic keep;
+    bool have_positive = false;
+    bool have_nonnegative = false;
+    for (auto arg : x.get_args()) {
+        auto newarg = apply(arg);
+        if (is_true(is_positive(*newarg, assumptions_))) {
+            keep.push_back(newarg);
+            have_positive = true;
+        } else if (is_true(is_nonnegative(*newarg, assumptions_))) {
+            keep.push_back(newarg);
+            have_nonnegative = true;
+        } else if (is_true(is_negative(*newarg, assumptions_))) {
+            negative.push_back(newarg);
+        } else if (is_true(is_nonpositive(*newarg, assumptions_))) {
+            nonpositive.push_back(newarg);
+        } else {
+            keep.push_back(newarg);
+        }
+    }
+    if (not have_positive and not nonpositive.empty()) {
+        std::copy(nonpositive.begin(), nonpositive.end(),
+                  std::back_inserter(keep));
+    }
+    if (not have_nonnegative and not have_positive and not negative.empty()) {
+        std::copy(negative.begin(), negative.end(), std::back_inserter(keep));
+    }
+
+    result_ = max(keep);
+}
+
+void RefineVisitor::bvisit(const Min &x)
+{
+    // negative < nonnegative
+    // nonpositive and negative < positive
+    vec_basic nonnegative;
+    vec_basic positive;
+    vec_basic keep;
+    bool have_negative = false;
+    bool have_nonpositive = false;
+    for (auto arg : x.get_args()) {
+        auto newarg = apply(arg);
+        if (is_true(is_negative(*newarg, assumptions_))) {
+            keep.push_back(newarg);
+            have_negative = true;
+        } else if (is_true(is_nonpositive(*newarg, assumptions_))) {
```

Comment:
Can this be written as `is_false(is_negative())` and reuse `is_negative()`?

Reply:
Thanks @isuruf for reviewing.   Yes, it should be possible.

Reply:
This turns out to be problematic. Both `is_false(is_negative(expr))` and `is_false(is_nonnegative(expr))` can be true in the case where `expr` is non-real. I can go ahead with this change, but then I would need to check `is_real` first.

The option to this is to let `is_negative` etc throw for non-real numbers. My current idea for all `is_` has been to allow all numeric input and throw for `Sets`, `Logic` etc.

Reply:
What does sympy do?

Reply:
Sympy does not throw for non-reals:

```python
>>> import sympy
>>> sympy.I.is_positive
False
```

So either what we have in the PR now or first check `is_real` then `is_negative` results can be inverted to get `is_nonnegative` for free. What do you think @isuruf?